File: smac_cli/live.py
"""Background WebSocket listener threads: `ChannelFeed` and `EventBell`.

Both connect via `websockets.sync.client` on a dedicated **daemon**
thread -- never the Textual event loop, which every blocking call in
this app is kept off of (see `smac_cli/app.py`'s module docstring). Each
delivers every decoded JSON payload to a caller-supplied callback
*directly on that background thread*; hopping back onto the UI thread
(`App.call_from_thread`) is the callback's own job, not this module's --
`smac_cli.live` intentionally never imports `textual`, so it stays
independently testable (a fake/echo WebSocket server, no Textual app
required) and reusable outside the TUI.

**Reconnection:** any failure at all -- a dropped connection, a refused
one, a bad `ws_url_provider` -- is treated identically: back off (1s,
doubling, capped at 30s, +0-25% jitter so many clients reconnecting at
once don't thunder-herd) and try again, forever, until `stop()`. The
`except Exception` in `_run` is deliberately broad: a reconnect loop
that can itself crash the daemon thread on some unanticipated failure
mode defeats the entire point of "auto-reconnect", and in the shell's
own unit tests (`tests/test_tui_shell.py`) the stub `FakeApi` doesn't
even implement `ws_channel_url`/`ws_events_url` -- the resulting
`AttributeError` is exactly the kind of failure this loop must absorb
rather than propagate.
"""

from __future__ import annotations

import json
import random
import threading
from typing import Any, Callable

from websockets.sync.client import ClientConnection, connect as ws_connect

_INITIAL_BACKOFF = 1.0
_MAX_BACKOFF = 30.0
_JITTER_FRACTION = 0.25
#: How long a single `recv()` blocks before looping back to recheck
#: `stop()` -- small enough that `stop()` takes effect promptly, large
#: enough to not busy-loop.
_RECV_POLL_SECONDS = 1.0


class _ReconnectingFeed:
    """Shared reconnect-loop machinery behind `ChannelFeed`/`EventBell`.

    `ws_url_provider` is called fresh on every (re)connect attempt --
    `SmacApi.ws_channel_url`/`.ws_events_url` mint a freshly-refreshed
    token per call (see `smac_cli/api.py`), so a token that went stale
    while this feed sat disconnected is never replayed.
    """

    def __init__(
        self,
        ws_url_provider: Callable[[], str],
        on_payload: Callable[[dict[str, Any]], None],
    ) -> None:
        self._ws_url_provider = ws_url_provider
        self._on_payload = on_payload
        self._stop_event = threading.Event()
        self._ws_lock = threading.Lock()
        self._ws: ClientConnection | None = None
        self._ever_connected = False
        self._thread = threading.Thread(target=self._run, daemon=True)

    def start(self) -> None:
        self._thread.start()

    def stop(self) -> None:
        """Signal the background thread to stop and unblock it promptly.

        Closing the live socket (if one is open right now) interrupts an
        in-progress `recv()` immediately rather than waiting out its
        poll timeout; a socket that's mid-(re)connect-attempt or asleep
        in the backoff wait notices `_stop_event` on its own within
        `_RECV_POLL_SECONDS`/one backoff tick. Does not join the thread
        -- it's a daemon, so process exit (or the next test) reaps it
        either way.
        """
        self._stop_event.set()
        with self._ws_lock:
            if self._ws is not None:
                try:
                    self._ws.close()
                except Exception:
                    pass

    def _on_disconnected(self) -> None:
        """Hook for a subclass to react to a connection drop. No-op by default."""

    def _on_reconnected(self) -> None:
        """Hook for a subclass to react to a successful *re*-connect (not
        the very first connect of this feed's lifetime). No-op by default."""

    def _run(self) -> None:
        backoff = _INITIAL_BACKOFF
        while not self._stop_event.is_set():
            try:
                url = self._ws_url_provider()
                with ws_connect(url, open_timeout=10) as ws:
                    with self._ws_lock:
                        self._ws = ws
                    if self._ever_connected:
                        self._on_reconnected()
                    self._ever_connected = True
                    backoff = _INITIAL_BACKOFF  # a healthy connect resets it
                    while not self._stop_event.is_set():
                        try:
                            raw = ws.recv(timeout=_RECV_POLL_SECONDS)
                        except TimeoutError:
                            continue
                        payload = json.loads(raw)
                        self._on_payload(payload)
            except Exception:
                if self._ever_connected:
                    self._on_disconnected()
            finally:
                with self._ws_lock:
                    self._ws = None

            if self._stop_event.is_set():
                break
            jitter = random.uniform(0, backoff * _JITTER_FRACTION)
            self._stop_event.wait(backoff + jitter)
            backoff = min(backoff * 2, _MAX_BACKOFF)


class ChannelFeed(_ReconnectingFeed):
    """The live feed for one channel's WebSocket (`/ws/.../channels/{id}`).

    Delivers every message payload broadcast on the channel to
    `on_payload`, plus two synthetic control payloads the server never
    sends -- `{"event": "disconnected"}` right when the connection
    drops, and `{"event": "reconnected"}` on the following successful
    reattach -- so the caller can show the spec's system lines and
    trigger a history refresh through the exact same `on_payload`
    conduit, keeping the constructor's 2-argument shape
    (`ChannelFeed(ws_url_provider, on_payload)`) intact rather than
    growing extra callback parameters for what is a rare event.
    """

    def _on_disconnected(self) -> None:
        self._on_payload({"event": "disconnected"})

    def _on_reconnected(self) -> None:
        self._on_payload({"event": "reconnected"})


class EventBell(_ReconnectingFeed):
    """The caller's private mention-events feed
    (`/ws/.../members/me/events`) -- delivers every `{"event":
    "mention", ...}` envelope to `on_event`. Reconnects silently (no
    synthetic control payloads): unlike a channel's message history,
    there's no "refresh" needed for a feed of point-in-time
    notifications, and the spec's reconnect UI is scoped to the channel
    feed only.
    """

File: smac_web/src/lib/live.ts
/**
 * The live layer (task-4 brief, web spec §3, constitution §5): a browser
 * `WebSocket` port of `smac_cli/live.py`'s `_ReconnectingFeed` machinery --
 * same reconnect discipline (exponential backoff, 1s doubling to a 30s
 * cap, +0-25% jitter so many clients reconnecting at once don't
 * thunder-herd), same "every dial gets a fresh token" contract (the
 * caller-supplied url provider is awaited fresh on every connect
 * attempt, never cached -- `api.wsChannelUrl`/`wsEventsUrl` each do a
 * `recoverWorkspaceSession()` internally before minting the URL, exactly
 * mirroring `SmacApi.ws_channel_url`/`.ws_events_url`'s docstrings).
 *
 * Two differences from the Python client, both deliberate simplifications
 * the task-4 brief calls for:
 *  - No synthetic `{"event": "disconnected"}"/"reconnected"}` control
 *    payloads. The TUI uses those to print system lines mid-feed; the web
 *    UI's reconnect UI need is fully covered by `onGap` (below) -- a
 *    disconnect itself needs no user-visible line, just the eventual
 *    catch-up.
 *  - `onGap` fires after EVERY successful (re)connect, including the very
 *    first one of this call's lifetime -- not just subsequent reconnects
 *    (contrast `smac_cli`'s `_on_reconnected`, which is a no-op on the
 *    first connect because the TUI already loads its own initial history
 *    before ever starting the feed). Here, `connectRoom` can be dialed
 *    the instant a room is entered, racing `workspace.tsx`'s own initial
 *    `refreshHistory()` fetch; firing `onGap` unconditionally on connect
 *    closes that race for free (`refreshUnreads`/`refreshHistory` are
 *    both idempotent GETs, so an extra call right after the store's own
 *    is harmless) rather than trying to detect and special-case "was this
 *    truly the first connect."
 */

import { wsChannelUrl, wsEventsUrl } from "./api";
import type { MentionEvent, MessagePayload } from "./types";

const INITIAL_BACKOFF_MS = 1000;
const MAX_BACKOFF_MS = 30_000;
const JITTER_FRACTION = 0.25;

export type Closeable = { close: () => void };

/**
 * Shared reconnect-loop machinery behind `connectRoom`/`connectBell`.
 * `urlProvider` is called fresh on every (re)connect attempt -- see this
 * module's docstring for why that matters. `onOpen` fires after every
 * successful connect (including the first); `onPayload` fires once per
 * decoded JSON frame.
 *
 * Teardown (`close()`) is synchronous and total: it flips `stopped`
 * BEFORE touching the socket or the pending-reconnect timer, so neither
 * the socket's own `onclose` handler nor an in-flight `urlProvider()`
 * promise can schedule a new reconnect after `close()` returns -- no
 * zombie timers, no dial landing after teardown.
 */
class ReconnectingSocket<TPayload> {
  private stopped = false;
  private socket: WebSocket | null = null;
  private backoff = INITIAL_BACKOFF_MS;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;

  constructor(
    private readonly urlProvider: () => Promise<string>,
    private readonly onPayload: (payload: TPayload) => void,
    private readonly onOpen?: () => void
  ) {}

  start(): void {
    void this.attemptConnect();
  }

  close(): void {
    this.stopped = true;
    if (this.reconnectTimer !== null) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    const socket = this.socket;
    this.socket = null;
    if (socket !== null) {
      // Detach the handlers first -- some WebSocket implementations
      // (including the test double) invoke `onclose` synchronously from
      // `close()`; `stopped` is already `true` above, so that handler's
      // own check is a belt-and-suspenders second guard, not the only one.
      socket.onopen = null;
      socket.onmessage = null;
      socket.onerror = null;
      socket.onclose = null;
      try {
        socket.close();
      } catch {
        // best-effort -- there is nothing left to do with a socket that
        // refuses to close cleanly during teardown.
      }
    }
  }

  private async attemptConnect(): Promise<void> {
    if (this.stopped) {
      return;
    }
    let url: string;
    try {
      url = await this.urlProvider();
    } catch {
      // A failed token mint/refresh is just another kind of connect
      // failure (mirrors `smac_cli/live.py`'s deliberately broad `except
      // Exception` -- the reconnect loop must never itself give up).
      this.scheduleReconnect();
      return;
    }
    if (this.stopped) {
      return;
    }

    let socket: WebSocket;
    try {
      socket = new WebSocket(url);
    } catch {
      // A malformed/rejected URL (or any other constructor failure) is
      // just another connect failure -- same broad-catch philosophy as
      // the `urlProvider()` rejection above, so a bad URL can never
      // crash the reconnect loop or surface as an unhandled rejection.
      this.scheduleReconnect();
      return;
    }
    this.socket = socket;

    socket.onopen = () => {
      if (this.stopped) return;
      this.backoff = INITIAL_BACKOFF_MS; // a healthy connect resets the backoff
      this.onOpen?.();
    };

    socket.onmessage = (event: MessageEvent) => {
      if (this.stopped) return;
      try {
        const data = JSON.parse(event.data as string) as TPayload;
        this.onPayload(data);
      } catch {
        // A malformed frame is dropped, not fatal -- the socket stays up.
      }
    };

    socket.onclose = () => {
      if (this.socket === socket) {
        this.socket = null;
      }
      if (this.stopped) return;
      this.scheduleReconnect();
    };

    socket.onerror = () => {
      // A browser WebSocket always follows "error" with "close" -- the
      // reconnect is scheduled from `onclose` above. This handler exists
      // solely so the "error" event is never left unhandled.
    };
  }

  private scheduleReconnect(): void {
    if (this.stopped) {
      return;
    }
    const jitter = Math.random() * this.backoff * JITTER_FRACTION;
    const delay = this.backoff + jitter;
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      void this.attemptConnect();
    }, delay);
    this.backoff = Math.min(this.backoff * 2, MAX_BACKOFF_MS);
  }
}

/**
 * The live feed for one channel (`/ws/.../channels/{channelId}`).
 * `onPayload` is called with every broadcast `MessagePayload`; `onGap`
 * fires after every successful (re)connect -- the caller's job (per the
 * task-4 brief) is to re-run `workspace.tsx`'s `refreshUnreads()` +
 * `refreshHistory()` from there, the "catch-up-then-live" discipline
 * (constitution §5) applied to a channel that may have missed messages
 * while its socket was down (or not yet open).
 */
export function connectRoom(
  channelId: string,
  onPayload: (payload: MessagePayload) => void,
  onGap: () => void
): Closeable {
  const socket = new ReconnectingSocket<MessagePayload>(
    () => wsChannelUrl(channelId),
    onPayload,
    onGap
  );
  socket.start();
  return { close: () => socket.close() };
}

/**
 * The caller's private mention-events feed
 * (`/ws/.../members/me/events`) -- delivers every `MentionEvent` to
 * `onEvent`. No gap callback: unlike a channel's message history, a feed
 * of point-in-time bell notifications has nothing to "catch up" on a
 * missed mention that already happened is simply gone -- there is
 * nothing to re-fetch (mirrors `smac_cli/live.py`'s `EventBell`, which reconnects
 * silently for the same reason).
 */
export function connectBell(onEvent: (event: MentionEvent) => void): Closeable {
  const socket = new ReconnectingSocket<MentionEvent>(() => wsEventsUrl(), onEvent);
  socket.start();
  return { close: () => socket.close() };
}
